### PR TITLE
:sparkles: Image lookup

### PR DESCRIPTION
**Pull request by avorima**
**What is the purpose of this pull request/Why do we need it?**

Implement image lookup at reconcile time according to label and name selectors. This allows to update the Kubernetes version only by modifying the CAPI resources instead of updating the image ID.

**Issue #, if available:**

**Description of changes:**

**Special notes for your reviewer:**

I added the following label to my images: `clusterapi=ionoscloud`.

Tested by deploying the new template and updating `KubeadmControlPlane` and `MachineDeployment` versions (1.28.7 -> 1.29.2).

```
kubectl patch kcp test-control-plane --type merge -p '{"spec":{"template":{"spec":{"version":"1.29.2"}}}}'
kubectl patch md test-worker --type merge -p '{"spec":{"version":"1.29.2"}}'
```

The PR is big as it is, so I'll look into an e2e test as a follow-up.

**Checklist:**
- [x] Documentation updated
- [x] Unit Tests added
- [ ] E2E Tests added
- [x] Includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)

### CODE REVIEW

File: docs/custom-image.md [hunk 1]
```diff
@@ -89,3 +89,22 @@ Now, you can copy the ID of your image and set it as the `IONOSCLOUD_MACHINE_IMA
...
 > Please ensure to update the KUBERNETES_VERSION in your environment file (envfile) if it changes.
+
+### Enabling image lookup
+
+The provider is able to look up images by label and name for `IonosCloudMachine` resources that make use of an image
+selector.
+By default the Kubernetes version of the parent Machine is used, so it's safe to reuse label keys and values for images
+that contain the version in their name.
+
+Currently it's only possible to label images using the REST API:
```
Comment (jriedel-ionos) [inline docs/custom-image.md:100]:
```suggestion
By default, the Kubernetes version of the parent Machine is used, so it's safe to reuse label keys and values for images
that contain the version in their name.

Currently, it's only possible to label images using the REST API:
```

**Review by jriedel-ionos — COMMENTED** [review]
and another template to keep in sync :heart_eyes_cat:

File: api/v1alpha1/ionoscloudmachine_types.go [hunk 2]
```diff
@@ -210,15 +210,36 @@ type Volume struct {
...
+	//+kubebuilder:validation:XValidation:rule="self.id != '' || has(self.selector)",message="must provide either id or selector"
 	Image *ImageSpec `json:"image"`
 }
 
 // ImageSpec defines the image to use for the VM.
 type ImageSpec struct {
-	// ID is the ID of the image to use for the VM.
+	// ID is the ID of the image to use for the VM. Has precedence over selector.
+	//
 	//+kubebuilder:validation:MinLength=1
```
Comment (lubedacht) [inline api/v1alpha1/ionoscloudmachine_types.go:221]: This validation can be removed because the default value is empty string and anything else has a min length of 1

File: api/v1alpha1/ionoscloudmachine_types.go [hunk 3]
```diff
@@ -210,15 +210,36 @@ type Volume struct {
...
+	// Only images in the IonosCloudCluster's location are considered.
+	//
+	//+optional
+	Selector *ImageSelector `json:"selector,omitempty"`
+}
+
+// ImageSelector defines label selectors for looking up images.
+type ImageSelector struct {
+	// MatchLabels is a map of key/value pairs.
+	MatchLabels map[string]string `json:"matchLabels"`
```
Comment (lubedacht) [inline api/v1alpha1/ionoscloudmachine_types.go:235]: Does a validation of at least one label in the map make sense?

File: api/v1alpha1/ionoscloudmachine_types_test.go [hunk 4]
```diff
@@ -311,16 +311,33 @@ var _ = Describe("IonosCloudMachine Tests", func() {
...
+					m.Spec.Disk.Image.ID = ""
+					m.Spec.Disk.Image.Selector = &ImageSelector{
+						MatchLabels: map[string]string{
+							"foo": "bar",
+						},
+					}
+					Expect(k8sClient.Create(context.Background(), m)).To(Succeed())
+					Expect(m.Spec.Disk.Image.Selector.UseMachineVersion).ToNot(BeNil())
+					Expect(*m.Spec.Disk.Image.Selector.UseMachineVersion).To(BeTrue())
+				})
```
Comment (lubedacht) [inline api/v1alpha1/ionoscloudmachine_types_test.go:340]: Test with `UseMachineVersion` nil missing

File: internal/service/cloud/image.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+func (s *Service) lookupImageID(ctx context.Context, ms *scope.Machine) (string, error) {
+	imageSpec := ms.IonosMachine.Spec.Disk.Image
+
+	if imageSpec.ID != "" {
+		return imageSpec.ID, nil
+	}
+
+	images, err := s.lookupImagesBySelector(ctx, ms.ClusterScope.Location(), imageSpec.Selector)
```
Comment (lubedacht) [inline internal/service/cloud/image.go:54]: The location is currently being reworked: https://github.com/ionos-cloud/cluster-api-provider-ionoscloud/pull/172.
So we either wait for it to be merged or we need a follow up here

File: internal/service/cloud/image.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+		images = filterImagesByName(images, version)
+	}
+
+	if len(images) == 0 {
+		return "", imageMatchError{selector: imageSpec.Selector}
+	}
+
+	if len(images) > 1 {
+		return "", imageMatchError{imageIDs: getImageIDs(images), selector: imageSpec.Selector}
+	}
```
Comment (lubedacht) [inline internal/service/cloud/image.go:74]: Up to you. Could be one `if` as getImageIDs would return empty slice

```suggestion
	if l := len(images); l == 0 || l > 1 {
		return "", imageMatchError{imageIDs: getImageIDs(images), selector: imageSpec.Selector}
	}
```

File: internal/service/cloud/image.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+	resourceLabels, err := s.ionosClient.ListLabels(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// mapping from image ID to labels
+	imageLabelMap := make(map[string]map[string]string)
+
+	for _, label := range resourceLabels {
+		if *label.Properties.ResourceType != "image" {
```
Comment (lubedacht) [inline internal/service/cloud/image.go:91]: We have been commonly using `ptr.Deref()`

Also the SDK provides `Get` functions which will never cause a panic if in the chain is any nil value

File: internal/service/cloud/image.go (hunk 5 above)
Comment (jriedel-ionos) [reply internal/service/cloud/image.go:54]: #172 merged

File: api/v1alpha1/ionoscloudmachine_types_test.go (hunk 4 above)
Comment (avorima) [reply api/v1alpha1/ionoscloudmachine_types_test.go:340]: do you mean i should verify that the default is set?

File: internal/service/cloud/image.go (hunk 7 above)
Comment (avorima) [reply internal/service/cloud/image.go:91]: i think that's a bit too defensive, but ok

File: api/v1alpha1/ionoscloudmachine_types_test.go (hunk 4 above)
Comment (lubedacht) [reply api/v1alpha1/ionoscloudmachine_types_test.go:340]: No I have been confusing myself. I thought you need to test the nil case because it is a pointer, but you have a default validation in there which defaults to `true`, therefore the default test case is enough that you already provided

**Review by lubedacht — APPROVED** [review]
